src/core/experimental/memory/memcpy.d

```diff
@@ -0,0 +1,401 @@
+/**
+ * Pure D replacement of the C Standard Library memcpy().
+ * There is an idiomatic-D interface, memcpy(), which is split into 3 overloads.
+ * One taking static types, one dynamic arrays and one static arrays.
+ * Also, there is available a C-like interface, the Dmemcpy() (which is named Dmemcpy
+ * for disambiguation with the C memcpy() a _similar_ interface) that
+ * is the classic (void*, void*, size_t) interface.
+ * N.B.: Both the memcpy() here and Dmemcpy() return nothing, contrary to the C Standard
+ * Library version.
+ * Source: $(DRUNTIMESRC core/experimental/memory/memcpy.d)
+ */
+module core.experimental.memory.memcpy;
+
+import core.internal.traits : isArray;
+
+/**
+ * Handle Static Types
+ * N.B.: No need for more sophisticated code for static types. The compiler
+ * knows better how to handle them in every target case.
+ *
+ * Params:
+ *  dst = Reference to memory destination to copy bytes to.
+ *  src = Reference to memory source to copy bytes from.
+ */
+pragma(inline, true)
+void memcpy(T)(ref T dst, ref const T src)
+if (!isArray!T)
```

Comment:
Why not use this as the user facing API and delegate with static if? Reduces the exposed interfaces -> less confusion.

Reply:
Because for me it's more confusing with one interface. I'll change it.

Reply:
I have a problem that I think I had when I had tried to join in one interface all of them. Basically, somehow the arrays get confused. e.g. if I have this code:

```
void memcpy(T)(ref T dst, ref const T src) nothrow @nogc
{
    static if (!isArray!T)
    {
        dst = src;
    }
    else
    {
        void memcpyArray(T)(ref T[] dst, ref const T[] src) nothrow @nogc
        {
            assert(dst.length == src.length);
            void* d = cast(void*) dst.ptr;
            const(void)* s = cast(const(void)*) src.ptr;
            size_t n = dst.length * typeof(dst[0]).sizeof;
            // Assume that there is no overlap.
            Dmemcpy(d, s, n);
        }

        static if (isDynamicArray!T)
        {
            memcpyArray(dst, src);
        }
        else
        {
            memcpyArray(dst[0 .. $], src[0 .. $]);
        }
    }
}
```